Importing Required Libraries


In [1]:
import pandas as pd
import numpy as np
import ast
import json
import datetime


In [2]:
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import plotly.io as pio

pio.renderers.default = "vscode"

Merging Tables & Initial Exploration 


In [3]:
movies = pd.read_csv("../Data/tmdb_5000_movies.csv")
credits = pd.read_csv("../Data/tmdb_5000_credits.csv")
df=movies.merge(credits,left_on="id",right_on="movie_id")

In [4]:
df.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ...  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                    title_x vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                    title_y  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce W

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

In [6]:
df.isna().sum()

budget                     0
genres                     0
homepage                3091
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title_x                    0
vote_average               0
vote_count                 0
movie_id                   0
title_y                    0
cast                       0
crew                       0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

Describing Data & Handling Inconsistent Values

In [8]:
df.describe()

budget             id   popularity       revenue      runtime  \
count  4.803000e+03    4803.000000  4803.000000  4.803000e+03  4801.000000   
mean   2.904504e+07   57165.484281    21.492301  8.226064e+07   106.875859   
std    4.072239e+07   88694.614033    31.816650  1.628571e+08    22.611935   
min    0.000000e+00       5.000000     0.000000  0.000000e+00     0.000000   
25%    7.900000e+05    9014.500000     4.668070  0.000000e+00    94.000000   
50%    1.500000e+07   14629.000000    12.921594  1.917000e+07   103.000000   
75%    4.000000e+07   58610.500000    28.313505  9.291719e+07   118.000000   
max    3.800000e+08  459488.000000   875.581305  2.787965e+09   338.000000   

       vote_average    vote_count       movie_id  
count   4803.000000   4803.000000    4803.000000  
mean       6.092172    690.217989   57165.484281  
std        1.194612   1234.585891   88694.614033  
min        0.000000      0.000000       5.000000  
25%        5.600000     54.000000    9014.500000  
50%        6.200000    235.000000   14629.000000  
75%        6.800000    737.000000   58610.500000  
max       10.000000  13752.000000  459488.000000

In [9]:
cols = ["budget", "popularity", "revenue", "runtime", "vote_average", "vote_count"]
df_zero = df[(df[cols] == 0).any(axis=1)]
df_zero = df_zero[["original_title"] + cols]
df_zero

original_title     budget  popularity    revenue  runtime  \
83                   The Lovers   27000000    2.418535          0    109.0   
135                 The Wolfman  150000000   21.214571          0    102.0   
265          The Cat in the Hat          0   18.251129          0     82.0   
309             Son of the Mask   84000000   17.815595          0     94.0   
321                The Campaign          0   16.460356  104907746     85.0   
...                         ...        ...         ...        ...      ...   
4797                     Cavite          0    0.022173          0     80.0   
4799                  Newlyweds       9000    0.642552          0     85.0   
4800  Signed, Sealed, Delivered          0    1.444476          0    120.0   
4801           Shanghai Calling          0    0.857008          0     98.0   
4802          My Date with Drew          0    1.929883          0     90.0   

      vote_average  vote_count  
83             4.8          34  
135            5.5         549  
265            4.9         366  
309            3.6         338  
321            5.6         578  
...            ...         ...  
4797           7.5           2  
4799           5.9           5  
4800           7.0           6  
4801           5.7           7  
4802           6.3          16  

[1576 rows x 7 columns]

In [10]:
cols = ["budget", "popularity", "revenue", "runtime", "vote_average", "vote_count"]

for col in cols:
    median_value = df[col].median()
    df.loc[df[col] == 0, col] = median_value

Data Cleaning

In [11]:
df.drop(["title_y","movie_id"],axis=1,inplace=True,errors="ignore")
df.columns=df.columns.str.strip()
pd.set_option("display.max_columns", None)
df.drop(columns=["homepage"], inplace=True ,errors="ignore")
df.dropna(subset=["release_date"])

budget                                             genres      id  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   19995   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...     285   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  206647   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   49026   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   49529   
...         ...                                                ...     ...   
4798     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...    9367   
4799       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   72766   
4800   15000000  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...  231617   
4801   15000000                                                 []  126186   
4802   15000000                [{"id": 99, "name": "Documentary"}]   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4798  [{"id": 5616, "name": "united states\u2013mexi...                es   
4799                                                 []                en   
4800  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4801                                                 []                en   
4802  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4798                               El Mariachi   
4799                                 Newlyweds   
4800                 Signed, Sealed, Delivered   
4801                          Shanghai Calling   
4802                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4798  El Mariachi just wants to play his guitar and ...   14.269792   
4799  A newlywed couple's honeymoon is upended by th...    0.642552   
4800  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4801  When ambitious New York attorney Sam is sent t...    0.857008   
4802  Ever since the second grade when he first saw ...    1.929883   

                                   production_companies  \
0     [{"name": "Ingenious Film Partners", "id": 289...   
1     [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2     [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3     [{"name": "Legendary Pictures", "id": 923}, {"...   
4           [{"name": "Walt Disney Pictures", "id": 2}]   
...                                                 ...   
4798           [{"name": "Columbia Pictures", "id": 5}]   
4799                                                 []   
4800  [{"name": "Front Street Pictures", "id": 3958}...   
4801                                                 []   
4802  [{"name": "rusty bear entertainment", 

Loading Data from JSON to Python

In [12]:
json_cols = ["genres", "keywords", "production_companies", "spoken_languages", "cast", "crew","production_countries"]
for col in json_cols:
    df[col]=df[col].apply(lambda x:ast.literal_eval(x) if isinstance(x,str) else x)

In [13]:
df["Genres_names"]=df["genres"].apply(lambda x:[d["name"]for d in x])


In [14]:
df["keywords"] = df["keywords"].apply(lambda x: [d["name"] for d in x])


In [15]:
df["production_Countries"]=df["production_countries"].apply(lambda x:[d["name"]for d in x])


In [16]:
df["Main_actor"]=df["cast"].apply(lambda x: x[0]["name"] if len(x)>0 else None)

In [17]:
df["All_cast"] = df["cast"].apply(lambda x:[d["name"]for d in x[:]])


In [18]:
df["Director"]=df["crew"].apply(lambda x:next((d["name"]for d in x if d["job"]=="Director"),None))

In [19]:
df["languages"]=df["spoken_languages"].apply(lambda x :[d["name"]for d in x])


In [20]:
df["main_production_company"] = df["production_companies"].apply(lambda x:[d["name"]for d in x])


Feature Engineering


In [21]:
df["release_date"] =pd.to_datetime(df["release_date"],errors="coerce")
df["release_year"] = df["release_date"].dt.year
df["release_month"] = df["release_date"].dt.month



In [22]:
df["release_year"] = df["release_year"].astype("Int64")
df["release_month"] = df["release_month"].astype("Int64")

In [23]:
df.isna().sum()

budget                       0
genres                       0
id                           0
keywords                     0
original_language            0
original_title               0
overview                     3
popularity                   0
production_companies         0
production_countries         0
release_date                 1
revenue                      0
runtime                      2
spoken_languages             0
status                       0
tagline                    844
title_x                      0
vote_average                 0
vote_count                   0
cast                         0
crew                         0
Genres_names                 0
production_Countries         0
Main_actor                  43
All_cast                     0
Director                    30
languages                    0
main_production_company      0
release_year                 1
release_month                1
dtype: int64

Filling Na

In [24]:
df["overview"]=df["overview"].fillna("No overview available")
df["Main_actor"] = df["Main_actor"].fillna("Unknown Actor")
df["Director"] = df["Director"].fillna("Unknown Director")
df["tagline"] = df["tagline"].fillna("")
df["budget"] = df["budget"].fillna(0)
df["revenue"] = df["revenue"].fillna(0)
df.to_csv("../Data/movies_clean.csv", index=False)


* Advanced Insights :

Revenue & Budget insights:


Profit Insights:

In [25]:
# Top10 movies by profit

df["profit"] = df["revenue"] - df["budget"]

top10_movies_profit = df.sort_values("profit", ascending=False).head(10)



# Top10 genres by profit

df_exploded = df.explode("Genres_names")
avg_profit_genre = (
    df_exploded.groupby("Genres_names")["profit"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)
# Runtime Stats
runtime_stats = df[df["runtime"].notna()].copy()
runtime_stats["runtime_category"] = pd.cut(
    runtime_stats["runtime"],
    bins=[0, 100, 140, 300],
    labels=["Short", "Medium", "Long"]
)

# Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("Top 10 Movies by Profit", "Top 10 Genres by Average Profit","Runtime vs Profit","Profit vs Vote Average")
)
# Subplot 1: Top10 movies by profit
fig.add_trace(
    go.Bar(x=top10_movies_profit["title_x"], y=top10_movies_profit["profit"], marker_color="teal"),
    row=1, col=1
)

# Subplot 2: Top10 genres by profit

fig.add_trace(
    go.Bar(x=avg_profit_genre.index, y=avg_profit_genre.values, marker_color="steelblue"),
    row=1, col=2
)

# Subplot 3: Runtime vs Profit 
fig.add_trace(
    go.Box(
        x=runtime_stats["runtime_category"],
        y=runtime_stats["profit"],
        marker_color="royalblue",
        name="Profit Distribution"
    ),
    row=2, col=1
)



# Subplot 4: Profit vs Vote Average
fig.add_trace(
    go.Scatter(
        x=runtime_stats["vote_average"],
        y=runtime_stats["profit"],
        mode="markers",
        marker=dict(color="purple"),
        name="Profit vs Vote Average"
    ),
    row=2, col=2
)



fig.update_layout(
    height=1000, width=1000,
    title_text="Profit Insights",
    font=dict(size=12, color="white"),
    margin=dict(t=80, b=80, l=60, r=60),
    template="plotly_dark"
)

fig.update_xaxes(tickangle=-30)

fig.show()

In [26]:

financial_df = ["popularity","budget","revenue","profit","vote_count","runtime"]
corr_matrix = df[financial_df].corr()

# 1) Top 10 Movies by Revenue
top10_revenue = df.sort_values("revenue", ascending=False).head(10)

# 2) Top 10 Genres by Profit
df["Genres_combo"] = df["Genres_names"].apply(lambda x: " | ".join(x) if isinstance(x, list) else x)
top10_genres_profit = (
    df.groupby("Genres_combo")["profit"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 3) Average Revenue & Budget by Genre
df_exploded = df.explode("Genres_names")
avg_rev_budget_genre = (
    df_exploded.groupby("Genres_names")[["revenue", "budget"]]
    .mean()
    .sort_values("revenue", ascending=False)
    .head(10)
)

# Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Top 10 Movies by Revenue",
        "Top 10 Genres combination by Average Profit",
        "Top 10 Genres by Average Revenue & Budget",
        "Financial Metrics Correlation Heatmap"
    )
)

# Subplot 1: Top 10 Movies by Revenue (Revenue + Budget)
fig.add_trace(
    go.Bar(x=top10_revenue["title_x"], y=top10_revenue["revenue"], name="Revenue", marker_color="seagreen"),
    row=1, col=1
)
fig.add_trace(
    go.Bar(x=top10_revenue["title_x"], y=top10_revenue["budget"], name="Budget", marker_color="orange"),
    row=1, col=1
)

# Subplot 2: Top 10 Genres by Profit
fig.add_trace(
    go.Bar(x=top10_genres_profit.index, y=top10_genres_profit.values, name="Profit", marker_color="mediumseagreen"),
    row=1, col=2
)

# Subplot 3: Average Revenue & Budget by Genre
fig.add_trace(
    go.Bar(x=avg_rev_budget_genre.index, y=avg_rev_budget_genre["revenue"], name="Revenue", marker_color="dodgerblue"),
    row=2, col=1
)
fig.add_trace(
    go.Bar(x=avg_rev_budget_genre.index, y=avg_rev_budget_genre["budget"], name="Budget", marker_color="orange"),
    row=2, col=1
)

# Subplot 4: Correlation Heatmap
fig.add_trace(
    go.Heatmap(
        z=corr_matrix.values,
        x=corr_matrix.columns,
        y=corr_matrix.columns,
        colorscale="RdBu",
        zmin=-1, zmax=1,
        text=corr_matrix.round(2).values,
        texttemplate="%{text}",
        hoverongaps=False
    ),
    row=2, col=2
)

fig.update_layout(
    height=1000, width=1000,
    title_text="Revenue, Profit & Budget Insights",
    font=dict(size=12, color="white"),
    margin=dict(t=80, b=80, l=60, r=60),
    barmode="group", template="plotly_dark"
)
fig.update_xaxes(tickangle=-30)

fig.show()

Popularity Insights:

In [27]:
df = df[(df["budget"] > 0) & (df["revenue"] > 0) & (df["runtime"] > 0) & (df["vote_count"] > 0)& (df["profit"] > 0)& (df["popularity"] > 0)]

# 1) Top 10 Movies by Popularity
top10_pop_movies = df.sort_values("popularity", ascending=False).head(10)

# 2) Top 10 Genres by Average Popularity
df_exploded = df.explode("Genres_names")
avg_pop_genre = (
    df_exploded.groupby("Genres_names")["popularity"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 3) Top Months by Average Profit
avg_profit_month = (
    df.groupby("release_month")["profit"]
    .mean()
    .sort_values(ascending=False)
)

# 4) Popularity Trend by Release Year
avg_pop_year = df.groupby("release_year")["popularity"].mean()

# Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Top 10 Movies by Popularity",
        "Top 10 Genres by Average Popularity",
        "Top Months by Average Profit",
        "Popularity Trend by Release Year"
    )
)

# Subplot 1: Top 10 Movies by Popularity
fig.add_trace(
    go.Bar(
        x=top10_pop_movies["title_x"],
        y=top10_pop_movies["popularity"],
        marker_color="purple",
        name="Movies"
    ),
    row=1, col=1
)

# Subplot 2: Top 10 Genres by Average Popularity
fig.add_trace(
    go.Bar(
        x=avg_pop_genre.index,
        y=avg_pop_genre.values,
        marker_color="darkcyan",
        name="Genres"
    ),
    row=1, col=2
)

# Subplot 3: Top Months by Average Profit
fig.add_trace(
    go.Bar(
        y=avg_profit_month.index,
        x=avg_profit_month.values,
        marker_color="seagreen",
        name="Avg Profit",
        orientation="h"
    ),
    row=2, col=1
)
fig.update_yaxes(title_text="Release Month", row=2, col=1)
fig.update_xaxes(title_text="Average Profit ($)", row=2, col=1)

# Subplot 4: Popularity Trend by Release Year
fig.add_trace(
    go.Scatter(
        x=avg_pop_year.index,
        y=avg_pop_year.values,
        mode="lines+markers",
        line=dict(color="orange"),
        name="Avg Popularity"
    ),
    row=2, col=2
)

fig.update_layout(
    height=900, width=1100,
    title_text="Popularity & Profit Insights",
    font=dict(size=12, color="white"),
    margin=dict(t=80, b=80, l=60, r=60),
    template="plotly_dark"
)

fig.update_xaxes(tickangle=-30)

fig.show()

Vote Insights:

In [28]:

# Top 10 movies by votes
top10_vote_movies = df.sort_values("vote_average", ascending=False).head(10)

# Top 10 Genres by votes
avg_vote_genre = (
    df_exploded.groupby("Genres_names")["vote_average"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# Vote Average vs Popularity Scatter
df_vote_pop=df[(df["vote_average"]>0)&(df["popularity"]>0)]

#  Vote Average Trend by Release Year
avg_vote_year = df.groupby("release_year")["vote_average"].mean()

fig=make_subplots(rows=2,cols=2,subplot_titles=("Top 10 Movies by Average Rating",
        "Top 10 Genres by Average Rating",
        "Vote Average vs Popularity Scatter",
        "Vote Average Trend by Release Year"
    )
)

# Subplots 2x2

# Subplot 1: Top 10 movies by votes

fig.add_trace(go.Bar(x=top10_vote_movies["title_x"], y=top10_vote_movies["vote_average"], marker_color="seagreen"),
    row=1, col=1
)

#subplot 2: Top 10 Genres by votes

fig.add_trace(go.Bar(x=avg_vote_genre.index, y=avg_vote_genre.values, marker_color="darkorange"),
    row=1, col=2
)

#subplot 3: Vote Average vs Popularity Scatter

fig.add_trace(
    go.Scatter(
        x=df_vote_pop["vote_average"], y=df_vote_pop["popularity"],
        mode="markers", marker=dict(color="purple", size=6),
        name="Movies", hovertext=df_vote_pop["title_x"], hoverinfo="text+x+y"
    ),
    row=2, col=1
)
fig.update_xaxes(title_text="vote average", row=2, col=1)
fig.update_yaxes(title_text="Popularity", row=2, col=1)

# Subplot 4: Trend by Year
fig.add_trace(
    go.Scatter(
        x=avg_vote_year.index, y=avg_vote_year.values,
        mode="lines+markers", line=dict(color="blue"),
        name="Avg Rating"
    ),
    row=2, col=2
)

fig.update_layout(
    height=900, width=1100,
    title_text="Vote Insights",
    font=dict(size=12, color="white"),
    margin=dict(t=80, b=80, l=60, r=60),
    template="plotly_dark"
)
fig.update_xaxes(tickangle=-30)

fig.show()



Runtime Insights:

In [29]:
df_runtime_votes = df[(df["runtime"] > 0) & (df["vote_count"] > 0)]
runtime_stats["Runtime_Category"] = pd.cut(
    runtime_stats["runtime"],
    bins=[0, 90, 120, 150, 180, float("inf")],
    labels=["Short (<90)", "Medium (90-120)", "Long (120-150)",
            "Very Long (150-180)", "Epic (>180)"]
)

runtime_counts = runtime_stats["Runtime_Category"].value_counts().sort_index()

avg_runtime_year = runtime_stats.groupby("release_year")["runtime"].mean()

avg_runtime_genre = (
    df_exploded.groupby("Genres_names")["runtime"]
    .mean()
    .sort_values(ascending=False)
)


#  Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Runtime vs Genre",
        "Movies Count by Runtime Category",
        "Runtime vs Votes",
        "Average Runtime Trend by Year"
    )
)
# Subplot 1: Runtime vs Genre
fig.add_trace(
    go.Bar(
        x=avg_runtime_genre.index,
        y=avg_runtime_genre.values,
        marker_color="royalblue",
        name="Avg Runtime"
    ),
    row=1, col=1
)





# Subplot 2: Movies Count by Runtime Category
fig.add_trace(
    go.Bar(
        x=runtime_counts.index,
        y=runtime_counts.values,
        marker_color="darkorange"
    ),
    row=1, col=2
)

# Subplot 3: Runtime vs Vote Count 
fig.add_trace(
    go.Scatter(
        x=df_runtime_votes["runtime"],
        y=df_runtime_votes["vote_count"],
        mode="markers",
        marker=dict(color="gold", size=6),
        name="Movies",
        hovertext=df_runtime_votes["title_x"],
        hoverinfo="text+x+y"
    ),
    row=2, col=1
)



#  Subplot 4: Trend by Year
fig.add_trace(
    go.Scatter(
        x=avg_runtime_year.index,
        y=avg_runtime_year.values,
        mode="lines+markers",
        line=dict(color="purple")
    ),
    row=2, col=2
)

fig.update_layout(
    height=1000, width=1000,
    title_text="Runtime Insights",
    font=dict(size=12, color="white"), 
    margin=dict(t=80, b=80, l=60, r=60),
    template="plotly_dark")         
fig.show()

Production Companies Insights:


In [30]:
df = df[(df["budget"] > 0) & (df["revenue"] > 0) & (df["runtime"] > 0) & (df["vote_count"] > 0)& (df["profit"] > 0)& (df["popularity"] > 0)]

df_exploded = df.explode("main_production_company")

#  Number of movies per company
company_counts = (
    df_exploded["main_production_company"]
    .value_counts()
    .head(10)
)

# Profit per company
profit_per_company = (
    df_exploded.groupby("main_production_company")["profit"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
)
# Average Runtime per Company
avg_runtime_company = (
    df_exploded.groupby("main_production_company")["runtime"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# Average rating per company
avg_vote_company = (
    df_exploded.groupby("main_production_company")["vote_average"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Top 10 Companies by Number of Movies",
        "Top 10 Companies by Profit",
        "Top 10 Companies by Avg Runtime",
        "Top 10 Companies by Avg Rating"
    )
)

# Subplot 1: Number of movies
fig.add_trace(
    go.Bar(y=company_counts.index, x=company_counts.values, marker_color="seagreen", name="Movies",orientation="h"),
    row=1, col=1
)

# Subplot 2: Top 10 companies by Profit
fig.add_trace(
    go.Bar(x=profit_per_company.index, y=profit_per_company.values, marker_color="dodgerblue", name="Revenue"),
    row=1, col=2
)

# Subplot 3: Average runtime per company
fig.add_trace(
    go.Bar(
        y=avg_runtime_company.index,          
        x=avg_runtime_company.values,         
        marker_color="mediumorchid",          
        name="Avg Runtime",
        orientation="h",                      
        text=avg_runtime_company.values,      
        textposition="outside"              
    ),
    row=2, col=1)

# Subplot 4: Average rating
fig.add_trace(
    go.Bar(x=avg_vote_company.index, y=avg_vote_company.values, marker_color="indianred", name="Rating"),
    row=2, col=2
)

fig.update_layout(
    height=900, width=1100,
    title_text="Production Companies Insights",
    barmode="group",
    template="plotly_dark",
    margin=dict(t=80, b=80, l=60, r=60)
)
fig.update_xaxes(tickangle=-30)

fig.show()

Language Insights


In [31]:
language_map = {
    "English": "English",
    "Español": "Spanish",
    "Français": "French",
    "Italiano": "Italian",
    "Deutsch": "German",
    "Türkçe": "Turkish",
    "ελληνικά": "Greek",
    "普通话": "Mandarin Chinese",
    "ภาษาไทย": "Thai",
    "Íslenska": "Icelandic",
    "Pусский": "Russian",
    "svenska": "Swedish",
    "Română": "Romanian",
    "日本語": "Japanese",
    "Latin": "Latin",
    "हिन्दी": "Hindi",
    "Português": "Portuguese",
    "فارسی": "Persian",
    "اردو": "Urdu",
    "العربية": "Arabic",
    "Český": "Czech",
    "广州话 / 廣州話": "Cantonese",
    "한국어/조선말": "Korean",
    "Norsk": "Norwegian",
    "தமிழ்": "Tamil",
    "עִבְרִית": "Hebrew",
    "Dansk": "Danish",
    "Nederlands": "Dutch",
    "Afrikaans": "Afrikaans",
    "Gaeilge": "Irish",
    "Somali": "Somali",
    "suomi": "Finnish",
    "Kiswahili": "Swahili",
    "български език": "Bulgarian",
    "Tiếng Việt": "Vietnamese",
    "Magyar": "Hungarian",
    "Український": "Ukrainian",
    "Esperanto": "Esperanto",
    "Polski": "Polish",
    "ਪੰਜਾਬੀ": "Punjabi",
    "Eesti": "Estonian",
    "shqip": "Albanian",
    "Srpski": "Serbian",
    "Bosanski": "Bosnian",
    "Hrvatski": "Croatian",
    "Slovenčina": "Slovak",
    "қазақ": "Kazakh",
    "తెలుగు": "Telugu",
    "Cymraeg": "Welsh",
    "Wolof": "Wolof",
    "isiZulu": "Zulu",
    "پښتو": "Pashto",
    "No Language": "Unknown",
    "Galego": "Galician",
    "ქართული": "Georgian",
    "বাংলা": "Bengali",
    "Català": "Catalan",
    "??????": "Unknown",
    "Bahasa indonesia": "Indonesian",
    "Bamanankan": "Bambara",
    "Slovenščina": "Slovenian",
    "nan": "Unknown",
    "": "Unknown"
}

In [32]:

df_exploded_lang = df.explode("languages")
df_exploded_lang["languages_clean"] = df_exploded_lang["languages"].map(language_map)

# 1) Top Languages by Number of Movies
top_languages_count = (
    df_exploded_lang["languages_clean"]
    .value_counts()
    .head(10)
)

# 2) Average Revenue per Language
avg_revenue_language = (
    df_exploded_lang.groupby("languages_clean")["revenue"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 3) Average Rating per Language
avg_rating_language = (
    df_exploded_lang.groupby("languages_clean")["vote_average"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 4) Average Profit per Language
avg_profit_language = (
    df_exploded_lang.groupby("languages_clean")["profit"]
    .mean()
    .sort_values(ascending=False)
    .head(5)
)

# Subplots 2×2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        "Top 10 Languages by Number of Movies",
        "Top 10 Languages by Avg Revenue",
        "Top 10 Languages by Avg Rating",
        "Top 5 Languages by Avg Profit"
    ),specs=[[{"type":"xy"}, {"type":"xy"}],
           [{"type":"xy"}, {"type":"domain"}]]
)

# Subplot 1: Number of Movies
fig.add_trace(
    go.Bar(
        y=top_languages_count.index,
        x=top_languages_count.values,
        orientation="h",
        marker_color="seagreen",
        text=top_languages_count.values,
        textposition="auto",
        name="Movies"
    ),
    row=1, col=1
)

# Subplot 2:  Revenue
fig.add_trace(
    go.Bar(
        x=avg_revenue_language.index,
        y=avg_revenue_language.values,
        marker_color="dodgerblue",
        text=avg_revenue_language.values.round(0),
        textposition="auto",
        name="Revenue"
    ),
    row=1, col=2
)

# Subplot 3: Rating
fig.add_trace(
    go.Scatter(
        x=avg_rating_language.index,
        y=avg_rating_language.values,
        mode="markers+lines",
        marker=dict(size=10, color="orchid"),
        name="Rating"
    ),
    row=2, col=1
)

# Subplot 4: Profit
fig.add_trace(
    go.Pie(
        labels=avg_profit_language.index,
        values=avg_profit_language.values,
        textinfo="label+percent",hole=0.5,
        marker=dict(colors=["gold","tomato","lightblue","lightgreen","violet"]),
        name="Profit"
    ),
    row=2, col=2
)

# Layout
fig.update_layout(
    height=900, width=1100,
    title_text="Languages Insights",
    template="plotly_dark",
    margin=dict(t=80, b=80, l=60, r=60)
)

fig.show()

Production Countries insights:

In [33]:

# 1) Top countries by number of movies
top_countries_count = df.explode("production_Countries")["production_Countries"].value_counts().head(10)

# 2) Avg revenue per country
avg_revenue_country = (
    df.explode("production_Countries")
    .groupby("production_Countries")["revenue"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 3) Avg rating per country
avg_rating_country = (
    df.explode("production_Countries")
    .groupby("production_Countries")["vote_average"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# Subplots 1x3 
fig = make_subplots(
    rows=1, cols=3,
    subplot_titles=(
        "Top Countries by Number of Movies",
        "Top Countries by Avg Revenue",
        "Top Countries by Avg Rating"
    )
)

# Subplot 1: Movies count
fig.add_trace(
    go.Bar(
        y=top_countries_count.index,
        x=top_countries_count.values,
        orientation="h",
        marker_color="limegreen",
        text=top_countries_count.values,
        textposition="auto",
        name="Movies"
    ),
    row=1, col=1
)

# Subplot 2: Revenue
fig.add_trace(
    go.Bar(
        x=avg_revenue_country.index,
        y=avg_revenue_country.values,
        marker_color="deepskyblue",
        text=avg_revenue_country.values.round(0),
        textposition="auto",
        name="Revenue"
    ),
    row=1, col=2
)

# Subplot 3: Rating
fig.add_trace(
    go.Scatter(
        x=avg_rating_country.index,
        y=avg_rating_country.values,
        mode="markers+lines",
        marker=dict(size=10, color="orchid"),
        name="Rating"
    ),
    row=1, col=3
)


fig.update_layout(
    height=600, width=1400,
    title_text="Production Countries Insights",
    template="plotly_dark",
    margin=dict(t=80, b=60, l=60, r=60)
)

fig.show()

Directors Insights:

In [34]:

df_director = df.copy()


df_director_grouped = df_director.groupby("Director")[["revenue","profit","vote_average"]].mean()

# Normalize 0-1
df_director_grouped["Revenue_Score"] = df_director_grouped["revenue"] / df_director_grouped["revenue"].max()
df_director_grouped["Profit_Score"] = df_director_grouped["profit"] / df_director_grouped["profit"].max()
df_director_grouped["Rating_Score"] = df_director_grouped["vote_average"] / df_director_grouped["vote_average"].max()

#  Composite Index
df_director_grouped["Performance_Index"] = (
    df_director_grouped["Revenue_Score"]*0.4 +
    df_director_grouped["Profit_Score"]*0.3 +
    df_director_grouped["Rating_Score"]*0.3
)

top_directors = df_director_grouped.nlargest(5, "Performance_Index").reset_index()



# Radar Chart
fig = go.Figure()
categories = ['Revenue', 'Profit', 'Rating']

for idx, row in top_directors.iterrows():
    values = [row['Revenue_Score'], row['Profit_Score'], row['Rating_Score']]
    fig.add_trace(go.Scatterpolar(
        r=values,
        theta=categories,
        fill='toself',
        name=row['Director'],
        line=dict(width=2)
    ))

fig.update_layout(
    polar=dict(radialaxis=dict(visible=True, range=[0,1])),
    showlegend=True,
    title="Multi-Dimensional Performance: Top 5 Directors (Radar Chart)",
    template="plotly_dark",
    height=700
)
fig.show()



Main Actors Insights:

In [35]:
# 1) Cast Size
df["cast_size"] = df["All_cast"].apply(lambda x: len(x) if isinstance(x, list) else 0)
df["cast_category"] = pd.cut(df["cast_size"], bins=[0,3,7,50], labels=["Small","Medium","Large"])

# 2) Top 10 Actors by Profit
df_exploded_cast = df.explode("All_cast")
top10_actors_profit = (
    df_exploded_cast.groupby("All_cast")["profit"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)

# 3) Top 10 Actors by Average Vote Rating
top10_actors_rating = (
    df_exploded_cast.groupby("All_cast")["vote_average"]
    .mean()
    .sort_values(ascending=False)
    .head(10)
)


# Subplots 2 x 2
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=("Cast Size vs Profit","Top 10 Actors by Profit","Top Actors by Rating","Profit vs Cast Size")
)

# Subplot 1: Cast Size vs Profit
fig.add_trace(
    go.Box(x=df["cast_category"], y=df["profit"], marker_color="royalblue"),
    row=1, col=1
)

# Subplot 2: Top 10 Actors by Profit
fig.add_trace(
    go.Bar(x=top10_actors_profit.index, y=top10_actors_profit.values, marker_color="seagreen"),
    row=1, col=2
)

# Subplot 3: Top 10 Actors by Average Rating
fig.add_trace(
    go.Bar(
        x=top10_actors_rating.index,
        y=top10_actors_rating.values,
        marker_color="violet",
        name="Top Actors by Rating"
    ),
    row=2, col=1
)


# Subplot 4: Profit vs Cast Size
fig.add_trace(
    go.Scatter(
        x=df["cast_size"],
        y=df["profit"],
        mode="markers",
        marker=dict(color="orange"),
        name="Profit vs Cast Size"
    ),
    row=2, col=2
)

fig.update_layout(
    height=1000, width=1100,
    title_text="All_cast Insights",
    template="plotly_dark",
    font=dict(size=12, color="white")
)

fig.show()

Production Companies Insights

In [36]:


df_companies_exploded = df.explode("main_production_company").dropna(subset=["main_production_company"])

company_stats = df_companies_exploded.groupby("main_production_company").agg({
    "vote_average": ["mean", "count", "max"],
    "profit": "mean",
    "revenue": "mean",
    "budget": "mean"
}).round(2)

company_stats.reset_index(inplace=True)
company_stats.columns = [
    "Company", "Avg_Rating", "Movie_Count", "Max_Rating",
    "Avg_Profit", "Avg_Revenue", "Avg_Budget"
]

company_stats["Avg_Profit/Movie_Count"] = company_stats["Avg_Profit"] / company_stats["Movie_Count"]

company_stats = company_stats[company_stats["Movie_Count"] >= 2].sort_values(
    "Avg_Profit/Movie_Count", ascending=False
)

top_companies = company_stats.head(10)

fig = px.bar(
    top_companies,
    x="Company",
    y="Avg_Profit/Movie_Count",
    color="Avg_Rating",
    hover_data=["Movie_Count", "Max_Rating", "Avg_Revenue", "Avg_Budget"],
    title="Production Company Performance: Avg Profit per Movie",
    labels={"Avg_Profit/Movie_Count": "Avg Profit per Movie", "Avg_Rating": "Average Rating"},
    color_continuous_scale="Viridis"
)

fig.update_layout(
    template="plotly_dark",
    height=700,
    width=1200,
    margin=dict(t=100, b=80, l=60, r=60),
    font=dict(size=14, color="white")
)

fig.show()


Conclusion & Recommendations



🎬 Insights from Correlation Matrix of Financial Metrics
The correlation matrix reveals key relationships between core financial and popularity metrics in movie data:
- Revenue vs Profit (0.97): Extremely strong positive correlation. Profit is almost directly tied to revenue, which is expected since profit = revenue − budget.
- Budget vs Revenue (0.73): Strong correlation. Higher-budget films tend to generate higher revenues, though not always proportionally.
- Popularity vs Vote Count (0.77): Strong correlation. Popular films attract more audience engagement and votes.
- Popularity vs Revenue (0.64): Moderate-to-strong correlation. Popularity contributes to revenue, but other factors (e.g., quality, marketing) also play a role.
- Budget vs Vote Count (0.59): Moderate correlation. Larger budgets often lead to wider exposure and more votes, but not as strongly as popularity does.

  Conclusion:
Financial success (revenue and profit) is closely linked to budget, while audience engagement (vote count) is strongly influenced by popularity. Popularity also affects revenue, but it's not the sole driver.


🎬 Insights from "Top 10 Genre Combinations by Average Profit"
The genre combination chart highlights which blends of genres yield the highest average profits:
- Genre combinations outperform single genres. All top entries are multi-genre blends, showing that mixing genres enhances audience appeal and profitability.
- Family-oriented and fantasy genres dominate. Combinations like Family | Fantasy | Adventure and Animation | Comedy | Family consistently rank high, indicating strong market demand for emotionally engaging, family-friendly content.
- Blending action, drama, and fantasy is lucrative. Combinations like Action | Drama | Fantasy and Action | Adventure | Fantasy | Science Fiction show that mixing intensity with imagination drives profitability.
- Science Fiction adds value when paired well. It appears in multiple top combinations, suggesting it boosts profit when combined with adventure or fantasy.
- Avoid genre over-specialization. No single-genre combinations appear in the top 10, implying that narrow genre focus may limit commercial success.

Conclusion:
Genre blending is a key driver of profitability. Investors and producers should prioritize multi-genre projects—especially those combining family, fantasy, adventure, and science fiction—to maximize return on investment.




🎬 Insights from Language &  Country

language:
- Average Rating by Language
- Languages like Telugu, Somali, and Tamil have the highest average ratings.
→ Some regional languages produce highly rated films, even if they’re less mainstream.
- Average Popularity by Language
- Languages like Română, Svenska, and Esperanto show high popularity scores.
→ Popularity doesn’t always align with rating — some languages attract attention even with average scores.
- Movie Count vs Revenue by Language
- English dominates in quantity, but Svenska, 廣東話, and ελληνικά lead in average revenue.

Country:
- The global film industry is heavily dominated by a few countries, especially the United States.
- The U.S. produces far more movies than any other country, shaping global trends and audience reach.
- This concentration highlights the imbalance in global production and the influence of major film hubs.

Recomendations:
Don’t rely only on volume — languages like Svenska and 廣東話 show strong revenue potential despite fewer films.




🎬 Key Insights from "Top Months by Average Profit"
1. Timing Directly Impacts Profitability
- Movies released in June (Month 6) achieve the highest average profit, followed closely by November (Month 11), December (Month 12) and May (Month 5).
- This suggests that release timing is a strategic financial decision, not just a scheduling one.
2. Summer and Holiday Seasons Are Highly Profitable
- The most profitable months align with summer vacation (May–July) and holiday season (December).
3. January Is a Low-Profit Month
- Movies released in January (Month 1) show the lowest average profit.



🎬 Popularity & Rating Insights
- Most movies have average ratings and low popularity.
- A few movies are very popular, but not always highly rated.
- Older movies (mid‑20th century) tend to have higher ratings.
- Recent movies show a decline in ratings compared to older ones.
- Most movies with high popularity don't always generate high profits, and some highly profitable films have relatively low popularity. This suggests that popularity alone isn't a reliable predictor of financial success — other factors like genre, timing, and production strategy play a major role.
Conclusion:
 Popularity ≠ Quality, and audience ratings have shifted over time


🎬 Insights from "Runtime"
- Genres with Longest Average Runtime
Historical and war movies tend to be the longest, while music and romance are shorter.
→ Genre affects runtime significantly.
- Movie Count by Runtime Category
Most movies fall into the "Medium" category (90–120 mins), with very few exceeding 180 mins.
→ Audiences and producers prefer mid-length films.
- Runtime vs Vote Count
There's no strong link between longer movies and higher vote counts. Most popular films are between 90–150 mins.
→ Longer runtime doesn’t guarantee more audience engagement.
- Average Runtime Over the Years
Runtime peaked in early cinema, dropped sharply, and stabilized around 100–120 mins in recent decades.
→ Modern films follow a consistent runtime trend.
- Movies with longer runtimes tend to have higher profit potential, especially with more extreme outliers.
→ Longer films might be riskier but can lead to bigger financial rewards.


.Recomendations
- While medium-length films are safer bets, longer movies show higher profit potential, especially with standout outliers.
- Focus on genres like drama and action that balance runtime and popularity.
